c/src/ml-api-service-extension.c

```diff
@@ -0,0 +1,1122 @@
+          }
+          break;
+        }
+        case ML_EXTENSION_TYPE_PIPELINE:
+        {
+          ml_extension_node_info_s *node_info;
+
+          g_mutex_lock (&mls->lock);
+          node_info = _ml_extension_node_info_get (ext, msg->name);
+          g_mutex_unlock (&mls->lock);
+
+          if (node_info && node_info->type == ML_EXTENSION_NODE_TYPE_INPUT) {
+            /* The input data will be released in the pipeline. */
+            status = ml_pipeline_src_input_data (node_info->handle, msg->input,
+                ML_PIPELINE_BUF_POLICY_AUTO_FREE);
+            msg->input = NULL;
+
+            if (status != ML_ERROR_NONE) {
+              _ml_error_report
+                  ("Failed to push input data into the pipeline in ml-service extension thread.");
+            }
+          } else {
+            _ml_error_report
+                ("Failed to push input data into the pipeline, cannot find input node '%s'.",
+                msg->name);
+          }
+          break;
+        }
+        default:
+          /* Unknown ml-service extension type, skip this. */
+          break;
+      }
+
+      _ml_extension_msg_free (msg);
+    }
+  }
+
+  return NULL;
+}
+
+/**
+ * @brief Wrapper to release tensors-info handle.
+ */
+static void
+_ml_extension_destroy_tensors_info (void *data)
+{
+  ml_tensors_info_h info = (ml_tensors_info_h) data;
+
+  if (info)
+    ml_tensors_info_destroy (info);
+}
+
+/**
+ * @brief Internal function to parse model path from json.
+ */
+static int
+_ml_extension_conf_parse_path (JsonNode * file_node, gchar ** path)
+{
+  guint i, n;
+
+  *path = NULL;
+
+  if (JSON_NODE_HOLDS_ARRAY (file_node)) {
+    JsonArray *array = json_node_get_array (file_node);
+    GString *val = g_string_new (NULL);
+
+    n = (array) ? json_array_get_length (array) : 0U;
+    for (i = 0; i < n; i++) {
+      const gchar *p = json_array_get_string_element (array, i);
+
+      g_string_append (val, p);
+      if (i < n - 1)
+        g_string_append (val, ",");
+    }
+
+    *path = g_string_free (val, FALSE);
+  } else {
+    *path = g_strdup (json_node_get_string (file_node));
+  }
+
+  return (*path != NULL) ? ML_ERROR_NONE : ML_ERROR_INVALID_PARAMETER;
+}
+
+/**
+ * @brief Internal function to parse tensors-info from json.
+ */
+static int
+_ml_extension_conf_parse_tensors_info (JsonNode * info_node,
+    ml_tensors_info_h * info_h)
+{
+  JsonArray *array = NULL;
+  JsonObject *object;
+  GstTensorsInfo info;
+  GstTensorInfo *_info;
+  const gchar *_str;
+  guint i;
+  int status;
+
+  gst_tensors_info_init (&info);
+
+  info.num_tensors = 1;
+  if (JSON_NODE_HOLDS_ARRAY (info_node)) {
+    array = json_node_get_array (info_node);
+    info.num_tensors = json_array_get_length (array);
+  }
+
+  for (i = 0; i < info.num_tensors; i++) {
+    _info = gst_tensors_info_get_nth_info (&info, i);
+
+    if (array)
+      object = json_array_get_object_element (array, i);
+    else
+      object = json_node_get_object (info_node);
```

Comment:
I think the object in line 375 is not always null.
Is this also true for the object in line 377?

Reply:
Possible null if json node is invalid, then it does not get the type or dimension. This will make failure tensor info.

Reply:
Thank you.